### PR TITLE
Simplify ZStream#combine To Not Use Queues

### LINKED ISSUE #1423: Simplify ZStream#combine to not use queues
With the new encoding, the queues are not necessary. The pulling function can work directly on the `InputStream` values from both sides.

### PR DESCRIPTION
**Pull request by adamgfraser**
Resolves #1423.

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -931,7 +931,7 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
       }
 
     for {
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: Slightly more efficient:

```scala
ZStream.fromInputStreamManaged {
  for {
    sides <- b.process
    result <- self.combine(...).process
  } yield result
}
```

File: streams/shared/src/main/scala/zio/stream/Take.scala [hunk 2]
```diff
@@ -60,6 +60,9 @@ object Take {
   final case class Value[A](value: A)       extends Take[Nothing, A]
   case object End                           extends Take[Nothing, Nothing]
 
+  final def fromInputStream[R, E, A](is: InputStream[R, E, A]): ZIO[R, Nothing, Take[E, A]] =
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/Take.scala]: 👍🏻

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 3]
```diff
@@ -1066,9 +1066,12 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
    * type with the specified mapping functions.
    */
   final def mergeWith[R1 <: R, E1 >: E, B, C](that: ZStream[R1, E1, B])(l: A => C, r: B => C): ZStream[R1, E1, C] = {
-    type Loser = Either[Fiber[Nothing, Take[E1, A]], Fiber[Nothing, Take[E1, B]]]
+    type Loser = Either[Fiber[Nothing, Take[E, A]], Fiber[Nothing, Take[E1, B]]]
 
-    def race(left: UIO[Take[E1, A]], right: UIO[Take[E1, B]]): UIO[(Take[E1, C], Loser)] =
+    def race(
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: I think we could probably get rid of `Take` entirely here, but we can do that in a followup.

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 4]
```diff
@@ -1406,25 +1409,23 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
...
       leftDone: Boolean,
       rightDone: Boolean,
-      q1: Queue[Take[E1, A]],
-      q2: Queue[Take[E1, B]]
+      left: InputStream[R, E, A],
+      right: InputStream[R1, E1, B]
     ): ZIO[R1, E1, ((Boolean, Boolean), Take[E1, C])] = {
-      val takeLeft: ZIO[R1, E1, Option[A]]  = if (leftDone) IO.succeed(None) else Take.option(q1.take)
-      val takeRight: ZIO[R1, E1, Option[B]] = if (rightDone) IO.succeed(None) else Take.option(q2.take)
+      val takeLeft: ZIO[R, E, Option[A]]    = if (leftDone) IO.succeed(None) else left.optional
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: `optional` is a nice touch here!

**Comment by adamgfraser** [comment]
@iravid Thanks for the comments! This is all set now.

**Pull request merged**